basicsr/archs/hifacegan_arch.py

```diff
@@ -0,0 +1,284 @@
+'''
+  HiFaceGAN_arch: Network definition of HiFaceGAN
+'''
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+# 20210519 [Lotayou]: Currently Pytorch official SyncBatchNorm only works under DDP training.
+# from basicsr.models.archs.sync_batchnorm import SynchronizedBatchNorm2d
+from basicsr.utils.registry import ARCH_REGISTRY
+from .hifacegan_util import BaseNetwork, LIPEncoder, SPADEResnetBlock, get_nonspade_norm_layer
+
+
+@ARCH_REGISTRY.register()
+class SPADEGenerator(BaseNetwork):
+    """Generator with SPADEResBlock"""
+
+    def __init__(self,
+                 num_in_ch=3,
+                 num_feat=64,
+                 use_vae=False,
+                 z_dim=256,
+                 num_upsampling_layers='normal',
+                 crop_size=512,
+                 norm_G='spectralspadesyncbatch3x3',
+                 is_train=True,
+                 init_train_phase=3):  # progressive training disabled
+        super().__init__()
+        self.nf = num_feat
+        self.input_nc = num_in_ch
+        self.is_train = is_train
+        self.train_phase = init_train_phase
+
+        if num_upsampling_layers == 'normal':
+            num_up_layers = 5
+        elif num_upsampling_layers == 'more':
+            num_up_layers = 6
+        elif num_upsampling_layers == 'most':
+            num_up_layers = 7
+        else:
+            raise ValueError(f'opt.num_upsampling_layers [{num_upsampling_layers}] not recognized')
+
+        # 20200211 Yang Lingbo with respect to phase
+        self.scale_ratio = num_up_layers
+        self.sw = crop_size // (2**num_up_layers)
+        self.sh = self.sw  # 20210519: By default use square image, aspect_ratio = 1.0
+
+        if use_vae:
+            # In case of VAE, we will sample from random z vector
+            self.fc = nn.Linear(z_dim, 16 * self.nf * self.sw * self.sh)
+        else:
+            # Otherwise, we make the network deterministic by starting with
+            # downsampled segmentation map instead of random z
+            self.fc = nn.Conv2d(num_in_ch, 16 * self.nf, 3, padding=1)
+
+        self.head_0 = SPADEResnetBlock(16 * self.nf, 16 * self.nf, norm_G)
+
+        self.G_middle_0 = SPADEResnetBlock(16 * self.nf, 16 * self.nf, norm_G)
```

Comment:
name convention, maybe g_middle_0

Reply:
Sure. Note that after this change the old checkpoints cannot be loaded due to key error. I'll provide new checkpoints after joining Tencent.